src/pointer-analysis/value_set.cpp

```diff
@@ -569,12 +569,34 @@ void value_sett::get_value_set_rec(
     // Find out what the pointer operand points at, and suck that data into
     // new object maps.
     object_mapt op0_set;
-    if(!is_pointer_type(op1))
-      get_value_set_rec(op0, op0_set, "", op0->type, false);
-
     object_mapt op1_set;
-    if(!is_pointer_type(op0))
-      get_value_set_rec(op1, op1_set, "", op1->type, false);
+    get_value_set_rec(op0, op0_set, "", (op0)->type, false);
+    get_value_set_rec(op1, op1_set, "", (op1)->type, false);
+
+    /*
+     * The main idea here is to be able to detect if any address might be valid at all.
+     * If any side of the pointer arithmetic is not a valid-object, then we can
+     * assume that it is not the pointer operand.
+     *
+     * As of now, there is one KNOWNBUG (esbmc/github_659_2_true), which is triggered by
+     * a cast of the index as char ptr. This makes ESBMC think that both operands are
+     * valid, which is currently unsupported.
+     */
+    auto op_contains_valid_address = [this](value_sett::object_mapt o) -> bool {
+      for(const auto &it : o)
+      {
+        auto expr = object_numbering[it.first];
+        if(is_symbol2t(expr) || is_constant_string2t(expr))
+          return true;
+      }
+
+      return false;
+    };
+
+    if(!op_contains_valid_address(op0_set))
```

Comment:
I believe ignoring the value-set of an operand based on non-presence of something is not sound: in particular, in the KNOWNBUG case the value-set of a `mul2t` expression is being built. As there is no special handling for that (yet), an `unknown2t` expression is put in there. By virtue of being unknown, it could also point to a symbol or constant string...

Reply:
This is an issue of not being complete, right? The value-set not being handled properly by multiplication causes it to propagate unknown.

`unknown` by itself can be anything. So, it is sound to throw an error.

Reply:
> This is an issue of not being complete, right? The value-set not being handled properly by multiplication causes it to propagate unknown.

Yeah, `unknown` as a fallback is fine if we don't handle it properly.

> `unknown` by itself can be anything. So, it is sound to throw an error.

Not sure what you mean by "throw an error". Where is that done? If `unknown` is in one operands value-set, I guess the easiest sound thing to do would be to put `unknown` in the entire expression's value-set.

But that's not done at the moment: the `unknown` in either operand's value-set is ignored, if it does **also** not contain whatever constitutes a "valid address".

Reply:
> Not sure what you mean by "throw an error".

Sorry, I meant at the KNOWNBUG example.

> If unknown is in one operands value-set, I guess the easiest sound thing to do would be to put unknown in the entire expression's value-set.

If the operand is the pointer operand, then yes we should propagate the unknown (which is what happens). If both operands do/don't have any valid place to go, then it will be unknown.

> But that's not done at the moment: the unknown in either operand's value-set is ignored, if it does also not contain whatever constitutes a "valid address".

This will be an issue only if the value-set has failed before. Even NULL pointer is considered a "valid address". I would guess that the main limitation is that if you have something like: `(int)(&var) + (int*)(nondet_int()))` or `nondet_int() + (int*)(nondet_int()))`  which will/should only propagate unknown.

Reply:
Maybe we're talking about different things? This code https://github.com/esbmc/esbmc/blob/45f57c96ad26ae4bb198c5a82d17b0bb11c66603/src/pointer-analysis/value_set.cpp#L596-L599 will remove any possibly existing `unknown` is the corresponding operand's value-set based on non-existance of something else in that set as determined by `op_contains_valid_address` ("valid address" above). That is what I perceive as unsound, since if that something else does not exist, the `unknown` will be dropped. Whether or not this also happens for the other operand is irrelevant for that statement.

Now, if `unknown` gets dropped here, then we're claiming that this operand does not refer to anything and correspondingly whatever it might have been referring to (`unknown`) will not be part of the formula, but instead will be encoded as an offset.

Reply:
> Maybe we're talking about different things? 

We aren't. But we still do not understand each other. `valid-address`, `value set` and `unknown` are too abstract. I will try to say what I understood from you:

I am assuming the following:
1. `unknown` is the same thing as nondet. Which, as you said, might be a valid address to any place.
2. Operations between two valid pointers does not make sense. It should be solved to an unknown (or error).
3. Operations between two invalid pointers does not make sense. It should be solved to an unknown (or error).
4. Valid address means a symbol or string. Which includes NULL pointer 

From those, 
1.  If we are trying to dereference it, then it will be a dereference error or a nondeterministic value that we cannot reason about it, e.g. `A = *(unknown_0); B = *(unknown_0); (A == B) fails`. *Is this what you mean?* 
2. This should be solved by a later PR, as it can be related to phi-operations.
3. This is already propagated as unknown.

Now, from what I understood from your comment is that you have one operator that the value-set know that it might be a valid-adress with a nondeterministic one is dropping the nondeterministic. Which I don't think it is, as doing the operation as an offset will cause it to be unknown.

The other flow is when both aren't valid, which will also propagate the unknown.

If this isn't what you meant. Then I will need a concrete example to be able to visualize.

Reply:
> `valid-address`, `value set` and `unknown` are too abstract.

- You introduced a small lambda and called it `op_contains_valid_address`. The condition `is_symbol2t(expr) || is_constant_string2t(expr)` in the check it performs is what I'm referring to, "valid address". Sure, this predicate is going to return `true` for value-sets containing `NULL` symbols, I know, and that's fine. But it will ignore `unknown2t` (and `invalid2t`) members of its argument, while I believe it should not.
- By `value-set (of an expression)` I am referring to what `get_value_set()` should compute in order for ESBMC to correctly reason about dereferencing. In particular, no possible pointed-to targets of the (however casted/arithmetically constructed) expression should be missed - unless `unknown2t` or `invalid2t` are elements of that set.
- `unknown` is a short-hand for `an unknown2t or invalid2t expression`.

> 1. If we are trying to dereference it, then it will be a dereference error or a nondeterministic value that we cannot reason about it, e.g. `A = *(unknown_0); B = *(unknown_0); (A == B) fails`. *Is this what you mean?*

Not quite, let's put this example in a `add2t` or `sub2t` context: `uintptr_t x, A = &x, B = (unknown_0); *(A+B) = 42`. This PR will call `op1_set.clear()` and then compute the value-set of `A+B` as the singleton `{ (x at an offset) }`. It is my understanding that instead it should either be the two-element set `{ (x at an offset), unknown2t }` or just `{ unknown2t }`. Is that understanding correct? (Related: the `plus-dlist.c` example in https://github.com/esbmc/esbmc/issues/677#issuecomment-1100605221 where expressions of the form `(&a+&b)-&a` result in addresses to unrelated to `a`.)

If so, then either the `op_contains_valid_address` predicate should not ignore `unknown2t` and `invalid2t` expressions or clearing the entire `op*_set` (which also removes these "placeholders" `unknown2t` and `invalid2t`) is too strong.

Reply:
> I am assuming the following:
> 1.    unknown is the same thing as nondet. Which, as you said, might be a valid address to any place.
> 2.    Operations between two valid pointers does not make sense. It should be solved to an unknown (or error).
> 3.    Operations between two invalid pointers does not make sense. It should be solved to an unknown (or error).
> 4.    Valid address means a symbol or string. Which includes NULL pointer

I believe 2. and 4. are not true.

2: see `plus-dlist.c` from the comment mentioned above or just consider `a+b` where `void *c = NULL; intptr_t b = (intptr_t)c`. The C standard defines `(void *)0` to be a null pointer constant. `NULL` is defined by the c2goto library exactly as that expression.

4: see the types mentioned in my 2nd remark to this PR, e.g. the constant array example in comment about the condition in `op_contains_valid_address` below.